source/opt/inline_pass.cpp

```diff
@@ -402,23 +468,117 @@ bool InlinePass::Inline(ir::Function* func) {
-  // We can only inline a function if it has blocks.
-  if (func->cbegin() == func->cend())
+bool InlinePass::hasMultipleReturns(ir::Function* func) {
+  bool seenReturn = false;
+  bool multipleReturns = false;
+  for (auto& blk : *func) {
+    auto lii = blk.cend();
+    lii--;
+    if (lii->opcode() == SpvOpReturn || lii->opcode() == SpvOpReturnValue) {
+      if (seenReturn) {
+        multipleReturns = true;
+        break;
+      }
+      seenReturn = true;
+    }
+  }
+  return multipleReturns;
+}
+
+void InlinePass::ComputeStructuredSuccessors(ir::Function* func) {
+  // If header, make merge block first successor.
+  for (auto& blk : *func) {
+    auto mii = blk.end();
+    mii--;
+    uint32_t mbid = 0;
+    if (mii != blk.begin()) {
+      mii--;
+      if (mii->opcode() == SpvOpLoopMerge)
+        mbid = mii->GetSingleWordOperand(kSpvLoopMergeMergeBlockId);
+      else if (mii->opcode() == SpvOpSelectionMerge)
+        mbid = mii->GetSingleWordOperand(kSpvSelectionMergeMergeBlockId);
+    }
+    if (mbid != 0)
+      block2structuredSuccs_[&blk].push_back(id2block_[mbid]);
+    // add true successors
+    blk.ForEachSuccessorLabel([&blk, this](uint32_t sbid) {
+      block2structuredSuccs_[&blk].push_back(id2block_[sbid]);
+    });
+  }
+}
+
+InlinePass::GetBlocksFunction InlinePass::SuccessorsFunction() {
+  return [this](const ir::BasicBlock* block) {
+    return &(block2structuredSuccs_[block]);
+  };
+}
+
+bool InlinePass::hasNoReturnInLoop(ir::Function* func) {
+  // If control not structured, do not do loop/return analysis
+  // TODO: Analyze returns in non-structured control flow
+  if (!module_->hasCapability(SpvCapabilityShader))
     return false;
-  // Do not inline functions with multiple returns
-  // TODO(greg-lunarg): Enable inlining if no return is in loop
-  int returnCnt = 0;
-  for (auto bi = func->cbegin(); bi != func->cend(); bi++) {
-    auto li = bi->cend();
-    li--;
-    if (li->opcode() == SpvOpReturn || li->opcode() == SpvOpReturnValue) {
-      if (returnCnt > 0)
-        return false;
-      returnCnt++;
+  // Compute structured block order. This order has the property
+  // that dominators are before all blocks they dominate and merge blocks
+  // are after all blocks that are control dependent on their header.
+  ComputeStructuredSuccessors(func);
+  auto ignore_block = [](cbb_ptr) {};
+  auto ignore_edge = [](cbb_ptr, cbb_ptr) {};
+  std::list<const ir::BasicBlock*> structuredOrder;
+  spvtools::CFA<ir::BasicBlock>::DepthFirstTraversal(
+    &*func->begin(), SuccessorsFunction(), ignore_block,
+    [&](cbb_ptr b) { structuredOrder.push_front(b); }, ignore_edge);
+  // Search for returns in loops. Only need to track outermost loop
+  bool return_in_loop = false;
+  uint32_t outerLoopMergeId = 0;
+  for (auto& blk : structuredOrder) {
+    // Exiting current outer loop
+    if (blk->id() == outerLoopMergeId)
+      outerLoopMergeId = 0;
+    // Return block
+    auto lii = blk->cend();
+    lii--;
```

Comment:
Prefer --lii

Side note: Seems it would be nice to have a .back() method on ir::BasicBlock.  :-)